### PR TITLE
Unify CLIs

### PR DESCRIPTION
**Pull request by stefmolin**
Changes:
- Created `numpydoc` entry point (replacing the hook's entry point).
- Moved CLI logic out of `__main__.py` and into a dedicated `cli.py` file.
- Refactored the CLI to use subparsers and treat each of the unique functionalities as a subcommand like `git`: `render`, `validate`, and `lint`.
- Updated `test_main.py` for changes and addition of `lint` option.
- Removed the `-m` option from the hook, so that usage is now from `numpydoc lint`.
- Updated the entry point for the pre-commit hook to `numpydoc lint`.
- Updated validation docs.
- Updated the `test.yml` workflow, but I didn't add a corresponding `lint` command as I don't fully understand what that part does.

Closes https://github.com/numpy/numpydoc/issues/467

### CODE REVIEW

**Review by jarrodmillman — APPROVED** [review]
Thanks!!

**Comment by jarrodmillman** [comment]
Stefan is back and will take a look at this soon. He is a little backlogged. Thanks for your patience.

**Comment by stefanv** [comment]
Thank you very much @stefmolin.

I see that I get a traceback when no import path is specified.

I think I suggested the `numpydoc` being an alias for `numpydoc render` before (bad Stéfan!), and now realize it is rather confusion. Perhaps we should let `numpydoc` (without arguments) print help? Or, at minimum, print a proper error message when `import_path` is None.

File: numpydoc/__main__.py [hunk 1]
```diff
@@ -2,55 +2,6 @@
...
-    )
-    args = ap.parse_args()
-
-    if args.validate:
-        exit_code = validate_object(args.import_path)
-    else:
-        exit_code = render_object(args.import_path, args.config)
-
-    sys.exit(exit_code)
+raise SystemExit(main())
```
Comment (stefanv) [inline numpydoc/__main__.py:7]: Neat, I never knew about SystemExit!

File: numpydoc/cli.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    subparsers = ap.add_subparsers(title="subcommands")
+
+    def _parse_config(s):
+        key, _, value = s.partition("=")
+        value = ast.literal_eval(value)
+        return key, value
+
+    render = subparsers.add_parser(
+        "render",
+        description="Test numpydoc docstring generation for a given object.",
```
Comment (stefanv) [inline numpydoc/cli.py:41]: Why does this "Test" docstring generation; doesn't it generate an expanded RST version of the docstring?

File: numpydoc/cli.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        type=_parse_config,
+        action="append",
+        help="key=val where val will be parsed by literal_eval, "
+        "e.g. -c use_plots=True. Multiple -c can be used.",
+    )
+    render.set_defaults(func=render_object)
+
+    validate = subparsers.add_parser(
+        "validate",
+        description="Validate the docstring with numpydoc.",
```
Comment (stefanv) [inline numpydoc/cli.py:57]: Can we clarify here the distinction between `validate` and `lint`. They do the same, except the one operates on a function and the other on a file? But `validate` doesn't allow disabling certain rules, which confused me.

File: numpydoc/cli.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        help="validate the object and report errors",
+    )
+    validate.add_argument("import_path", help="e.g. numpy.ndarray")
+    validate.set_defaults(func=validate_object)
+
+    lint_parser = validate_docstrings.get_parser(parent=subparsers)
```
Comment (stefanv) [inline numpydoc/cli.py:63]: How is the lint_parser hooked up? Why not the same way as render and validate?

Also, the lint docstring on my machine contains:

```
Check codes to ignore. Currently ignoring the following from /home/stefan/src/numpy/pyproject.toml: Values provided here will be
                        in addition to the above, unless an alternate config is provided.
```

Not quite right?

File: numpydoc/hooks/validate_docstrings.py [hunk 5]
```diff
@@ -357,10 +370,21 @@ def main(argv: Union[Sequence[str], None] = None) -> int:
         + "\n"
     )
 
-    parser = argparse.ArgumentParser(
-        description="Run numpydoc validation on files with option to ignore individual checks.",
-        formatter_class=argparse.RawTextHelpFormatter,
+    description = (
+        "Run numpydoc validation on files with option to ignore individual checks."
     )
+    if parent is None:
```
Comment (stefanv) [inline numpydoc/hooks/validate_docstrings.py:376]: I guess this is the part that confused me; why have two ways of adding the subparser? And, worse, without a parent automatically add lint subparser?

**Review by stefanv — COMMENTED** [review]
Looks good, I think the `lint` subparser perhaps needs to be hooked up more cleanly, but tested everything locally and it works. Thank you again!

File: numpydoc/cli.py (hunk 2 above)
Comment (stefmolin) [reply numpydoc/cli.py:41]: This actually comes from the original docstring for `render_object()`, which was the only thing called for the original CLI's behavior:

https://github.com/numpy/numpydoc/blob/3a21e5440c90a0f306c42fdc303083a12927969b/numpydoc/__main__.py#L14

File: numpydoc/cli.py (hunk 3 above)
Comment (stefmolin) [reply numpydoc/cli.py:57]: I think it works for non-functions as well. How about this?

> Validate an object's docstring conforms to the numpydoc standard.

File: numpydoc/cli.py (hunk 2 above)
Comment (stefanv) [reply numpydoc/cli.py:41]: :shrug: That's weird... but there are lots of weird things in this code :) Feel free to change whatever you see fit, or if it feels out of scope that's fine too.

File: numpydoc/cli.py (hunk 3 above)
Comment (stefanv) [reply numpydoc/cli.py:57]: "Validate that an object's docstring conforms to the numpydoc standard" :+1:

Comment (stefanv) [reply numpydoc/cli.py:57]: Or, "Determine (or check?) whether an object's docstring..."

Comment (stefanv) [reply numpydoc/cli.py:57]: Otherwise just: "Validate an object's docstring against the numpydoc standard"

Comment (stefanv) [reply numpydoc/cli.py:57]: I may prefer the last version, but anything gramatically sound is fine.

Comment (stefanv) [reply numpydoc/cli.py:57]: > I'm not sure what to do about the second part of your comment. If you run `numpydoc validate --help` you see what can be passed as arguments, and there aren't any for disabling rules.

This was more of a question: *should* there be the option of disabling certain rules? Why are these not exactly equivalent in their functionality?

Comment (stefmolin) [reply numpydoc/cli.py:57]: > > I'm not sure what to do about the second part of your comment. If you run `numpydoc validate --help` you see what can be passed as arguments, and there aren't any for disabling rules.
> 
> This was more of a question: _should_ there be the option of disabling certain rules? Why are these not exactly equivalent in their functionality?

While I agree this should eventually be the case, this definitely seems out of scope for this PR, but at least the separate subcommands will not draw attention to arguments that have no affect like before (see https://github.com/numpy/numpydoc/issues/459).

File: numpydoc/cli.py (hunk 4 above)
Comment (stefmolin) [reply numpydoc/cli.py:63]: There was a bug with this hack. I will move things around.

File: numpydoc/cli.py (hunk 2 above)
Comment (stefmolin) [reply numpydoc/cli.py:41]: I will use a version of what you put above. This PR has lots of potential for snowballing, and I want to keep it as focused as possible.

**Review by stefanv — APPROVED** [review]
Thanks @stefmolin! The new interface is great.

**Comment by stefanv** [comment]
@rossbar @jarrodmillman handing it over to you.

**Pull request merged**